### PR TITLE
refine semantic analaysis and fix bug with constants in filters

### LINKED ISSUE #2261: consts not carried into kernel.Filter
This fails because the const is not carried into the filter AST in kernel.Filter.
```
zql: const t = {a:string} ; is(t)

input: |
  {a:"hello"}
  {a:2}
  {a:"world"}
output: |
  {a:"hello"}
  {a:"world"}
```

### PR DESCRIPTION
**Pull request by mccanne**
This commit further develops the semantic analysis component of the
compiler.  This was needed to fix a problem where the lift-filter
optimization did not carry with it any Z constants that were defined
in the Z program causing such constant refs to be treated as field
references in the filters generated by the zbuf.Filter methods.

The fix is to fix the intermediate represenation of the Z program
to explicitly track the global scope of the Z program and initialize
constants in the runtime zctx before any filters are created.  Also,
when distributed workers are generated and use the AsProc() method
to zbuf.Filter, the AST returned here needs to include said constants
so this has been fixed.

In doing this refactor, we took one step toward the work needed
to split the parser AST and semantic DSL into logically distinct
components as outlined in issue #2163.

In the new model, the compiler stages are all abstracted into a single
entry point in the compiler.Runtime object.  The ast command has been
reworked to show the AST at ay stage of compilation (after parsing,
after semantic analysys, after optimization, after parallelization,
and so forth).  Having the taps into compiler.Runtime from the ast
command makes it very easy to see what the various stages do and whether
they are working correctly and it allowed us to recast the complex
tests for the parallelization logic into simple ztests that use the
ast command.

In a future PR, we will more cleanly separate the parser AST from
the kernel DSL AST.  For now, the two logically separate AST forms
share the same data structure in compiler/ast.

Fixes #2261

### CODE REVIEW

File: proc/groupby/groupby_test.go [hunk 1]
```diff
@@ -370,26 +380,28 @@ func TestGroupbyUnit(t *testing.T) {
...
+			setSortDir(runtime.Entry(), dir)
+
+			pctx := proctest.NewTestContext(zctx)
 			src := proctest.NewTestSource(inBatches)
-			gproc, err := proctest.CompileTestProcAST(astProc, tctx, src)
+			err = runtime.Compile(nil, pctx, []proc.Interface{src})
 			assert.NoError(t, err)
-			procTest := proctest.NewProcTest(gproc, tctx)
+			outputs := runtime.Outputs()
+			assert.Equal(t, 1, len(outputs))
```
Comment (nwt) [inline proc/groupby/groupby_test.go:400]:
```suggestion
			assert.Len(t, outputs, 1)
```

File: zfmt/canon.go [hunk 2]
```diff
@@ -306,6 +326,14 @@ func (c *canon) proc(p ast.Proc) {
 	}
 }
 
+func isTrue(e ast.Expression) bool {
+	if lit, ok := e.(*ast.Literal); ok {
+		ok = lit.Type == "bool" && lit.Value == "true"
+		return ok
```
Comment (nwt) [inline zfmt/canon.go:332]:
```suggestion
		return lit.Type == "bool" && lit.Value == "true"
```

**Review by nwt — COMMENTED** [review]
Always enjoy

File: compiler/compiler.go [hunk 3]
```diff
@@ -7,10 +7,124 @@ import (
...
+var _ zbuf.Filter = (*Runtime)(nil)
+
+type Runtime struct {
+	zctx    *resolver.Context
+	scope   *kernel.Scope
+	sem     *semantic.AST
+	outputs []proc.Interface
+}
+
+func New(zctx *resolver.Context, parserAST ast.Proc) (*Runtime, error) {
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/compiler.go:24]: Maybe the problem is with "runtime"?

File: compiler/kernel/proc.go [hunk 4]
```diff
@@ -325,21 +325,14 @@ func Compile(custom Hook, node ast.Proc, pctx *proc.Context, scope *Scope, paren
 		return nil, errors.New("no parents")
 	}
 	if scope == nil {
-		// Outermost caller should pass in global scope object.  If nil,
-		// we assume no global context and allocate a fresh, empty scope.
-		scope = newScope()
-		scope.Enter()
+		panic("no scope")
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/kernel/proc.go:328]: This is a bona fide bug and having the stack trace would be better no?  I could leave out the check and let it panic on the null pointer, but it shouldn't happen.

File: compiler/compiler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/compiler.go:24]: I tried changing it to "Runtime" and it doesn't read well to me.  You're creating a compiler that generates a runtime and when you're done you have a runtime object.  I kind of like just leaving it as New() but I understand your reaction.

**Review by nwt — APPROVED** [review]

File: compiler/compiler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwt) [reply compiler/compiler.go:24]: `New` works for me.

File: compiler/kernel/proc.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nwt) [reply compiler/kernel/proc.go]: Good point!

**Pull request merged**